# Some of these packages are better installed by biocmanager. make sure it is installed and up to date.

##### install biocmanager if not already
#### if (!require("BiocManager", quietly=F))
####   install.packages("BiocManager")
##### install txdb maker
#### BiocManager::install("txdbmaker")
##### and genomic features (not used here but in other scripts)
#### BiocManager::install("GenomicFeatures")

# load package
library(txdbmaker)
library(GenomicFeatures)
library(Rsamtools)

# setting the wd allows you to read/write to this dir without specifying full paths.
setwd('C:/Users/danie/OneDrive - Colostate/NIFA_PROJECT/Obj1/analysis/promoter_analysis/genome_gff/')
# its also good to do this before saving anything just in case you forgot to set it!

# Path to your genome FASTA
rpadi_genome = "R_padi_v2.fasta"

## make or read an index for that FASTA
## If it's not already indexed, you can do:
#indexFa(rpadi_genome) # creates an .fai

# In my case I have a fai that was generated by hisat2.
# load the FASTA index as a GRanges
faidx = scanFaIndex(rpadi_genome)

head(faidx)

## construct the chrominfo data.frame
chrominfo_df = data.frame(
  chrom = as.character(seqnames(faidx)),
  length = width(faidx),
  is_circular = rep(FALSE, length(faidx))  # for eukaryotes this will be false
)

head(chrominfo_df)


# load path to gff file
rpadi_gff = "C:/Users/danie/OneDrive - Colostate/NIFA_PROJECT/Obj1/analysis/promoter_analysis/genome_gff/Rpadi_v2.gff"


# generate tx database
rpadi_txdb = makeTxDbFromGFF(rpadi_gff, 
                dataSource = "AphidBase Annotation Version 2",
                organism = "Rhopalosiphum padi",
                taxonomyId = 40932,
                chrominfo = chrominfo_df
                )
# check database 
rpadi_txdb
# Db type: TxDb
# Supporting package: GenomicFeatures
# Data source: AphidBase Annotation Version 2
# Organism: Rhopalosiphum padi
# Taxonomy ID: 40932
# miRBase build ID: NA
# Genome: NA
# Nb of transcripts: 26535
# Db created by: txdbmaker package from Bioconductor
# Creation time: 2025-03-01 02:44:18 -0700 (Sat, 01 Mar 2025)
# txdbmaker version at creation time: 1.2.1
# RSQLite version at creation time: 2.3.9
# DBSCHEMAVERSION: 1.2

# we can save the database, but first set the wd to the dir of interest
setwd('C:/Users/danie/OneDrive - Colostate/NIFA_PROJECT/Obj1/analysis/promoter_analysis/genome_gff/')
saveDb(rpadi_txdb, file="rpadi_V2_txdb.sqlite")


